README.md
# Ethereum Alarm Clock Client


This is a client that can be used to monitor the alarm service for upcoming
scheduled calls and execute them when the appropriate block number is reached.

This is only compatable with the 0.6.0 release of alarm.

## Warning

This software should be considered alpha quality.  Please feel free to reach out to me with any issues you run into.

Some things that should probably be added before this is really ready for
public consumption:

* Claiming of scheduled calls.
* Conversion from RPC to IPC


## Installation

```bash
$ pip install ethereum-alarm-clock-client
```


## Usage

You can view all of the available commands:

```bash
$ eth_alarm
Usage: eth_alarm [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  addresses  List the addresses for different versions of...
  scheduler  Run the call scheduler.
```

Or to run the scheduler

```bash
$ eth_alarm scheduler
...
```

In order for the scheduler to function there must be a unlocked JSON-RPC server
running locally.


CHANGELOG
7.0.0
-----

- Switch to major version scheme as alarm version.
- Update for V7 alarm service

0.7.2
-----

- Add support for using ipc client.
- Update versions of rpc client.


0.7.1
-----

- Initial public release


tests/conftest.py
import pytest

from collections import defaultdict
import time


@pytest.fixture(autouse=True)
def alarm_client_logging_config(monkeypatch):
    monkeypatch.setenv('LOG_LEVEL', 'ERROR')


class MockBlockchainClient(object):
    def __init__(self, *blocks):
        self.blocks = list(blocks)
        if not self.blocks:
            self.mine()

    def get_block_by_number(self, block_number, full_transactions=False):
        try:
            int_block_number = int(block_number, 16)
            hex_block_number = block_number
        except (ValueError, TypeError):
            int_block_number = block_number
            hex_block_number = hex(block_number)

        return {
            'blockNumber': hex_block_number,
            'timestamp': hex(self.blocks[int_block_number - 1]),
        }

    def get_block_number(self):
        return len(self.blocks)

    def mine(self, timestamp=None):
        if timestamp is None:
            timestamp = time.time()
        self.blocks.append(int(timestamp))


@pytest.fixture()
def mock_blockchain_client_class():
    return MockBlockchainClient


@pytest.fixture()
def mock_blockchain_client(mock_blockchain_client_class):
    return mock_blockchain_client_class()


class MockLogger(object):
    def __init__(self):
        self.logs = defaultdict(list)

    def info(self, *args, **kwargs):
        self.logs['info'].append((args, kwargs))

    def warning(self, *args, **kwargs):
        self.logs['warning'].append((args, kwargs))

    def debug(self, *args, **kwargs):
        self.logs['debug'].append((args, kwargs))

    def error(self, *args, **kwargs):
        self.logs['error'].append((args, kwargs))


@pytest.fixture()
def mock_logger():
    return MockLogger()


class MaxWaitExceeded(Exception):
    pass


def _wait_till(condition, max_wait=15, message=None):
    start = time.time()
    while time.time() - start < max_wait:
        if condition():
            break
        time.sleep(0.1)
    else:
        if message is None:
            message = "Condition not met within {0} seconds".format(max_wait)
        raise MaxWaitExceeded(message)


@pytest.fixture()
def wait_till():
    return _wait_till


@pytest.fixture(scope="session")
def denoms():
    from ethereum.utils import denoms as ether_denoms
    return ether_denoms


tests/v6/conftest.py
import pytest


@pytest.fixture(scope="module")
def FutureBlockCall(contracts, deployed_contracts):
    from populus.contracts import (
        link_contract_dependency,
    )
    return link_contract_dependency(
        link_contract_dependency(contracts.FutureBlockCall, deployed_contracts.CallLib),
        deployed_contracts.AccountingLib,
    )


@pytest.fixture
def deploy_future_block_call(deploy_client, FutureBlockCall, deploy_coinbase):
    from populus.contracts import (
        deploy_contract,
    )
    from populus.utils import (
        get_contract_address_from_txn,
    )

    def _deploy_future_block_call(contract_function=None, scheduler_address=None,
                                  target_block=None, grace_period=255,
                                  required_gas=1000000, payment=1, donation=1,
                                  endowment=None, call_data="",
                                  require_depth=0, call_value=0):
        if endowment is None:
            endowment = deploy_client.get_max_gas() * deploy_client.get_gas_price() + payment + donation + call_value

        if target_block is None:
            target_block = deploy_client.get_block_number()

        if scheduler_address is None:
            scheduler_address = deploy_coinbase

        if contract_function is None:
            abi_signature = ""
            contract_address = scheduler_address
        else:
            abi_signature = contract_function.encoded_abi_signature
            contract_address = contract_function._contract._meta.address

        deploy_txn_hash = deploy_contract(
            deploy_client,
            FutureBlockCall,
            constructor_args=(
                scheduler_address,
                target_block,
                grace_period,
                contract_address,
                abi_signature,
                call_data,
                call_value,
                required_gas,
                require_depth,
                payment,
                donation,
            ),
            gas=int(deploy_client.get_max_gas() * 0.95),
            value=endowment,
        )

        call_address = get_contract_address_from_txn(deploy_client, deploy_txn_hash, 180)
        call = FutureBlockCall(call_address, deploy_client)
        return call
    return _deploy_future_block_call


@pytest.fixture()
def scheduled_call(deployed_contracts, deploy_future_block_call, deploy_client):
    call = deploy_future_block_call(
        deployed_contracts.TestCallExecution.setBool,
        target_block=deploy_client.get_block_number() + 400,
    )
    return call


@pytest.fixture()
def call_contract(scheduled_call, deployed_contracts, deploy_client):
    from eth_alarm_client.call_contract import CallContract
    from eth_alarm_client.scheduler import Scheduler

    scheduler = Scheduler(deployed_contracts.Scheduler)
    call_contract = CallContract(
        call_address=scheduled_call._meta.address,
        blockchain_client=deploy_client,
        block_sage=scheduler.block_sage,
    )
    return call_contract


@pytest.fixture(scope="module")
def CallLib(deployed_contracts):
    return deployed_contracts.CallLib


@pytest.fixture(scope="module")
def SchedulerLib(deployed_contracts):
    return deployed_contracts.SchedulerLib


@pytest.fixture(scope="module")
def get_call(SchedulerLib, FutureBlockCall, deploy_client):
    def _get_call(txn_hash):
        call_scheduled_logs = SchedulerLib.CallScheduled.get_transaction_logs(txn_hash)
        if not len(call_scheduled_logs) == 1:
            reject_logs = SchedulerLib.CallRejected.get_transaction_logs(txn_hash)
            if reject_logs:
                reject_data = SchedulerLib.CallRejected.get_log_data(reject_logs[0])
                raise ValueError("Call rejected: {0}".format(reject_data['reason']))
            raise ValueError("Call was not scheduled")
        call_scheduled_data = SchedulerLib.CallScheduled.get_log_data(call_scheduled_logs[0])

        call_address = call_scheduled_data['call_address']
        call = FutureBlockCall(call_address, deploy_client)
        return call
    return _get_call


#FUTURE_OFFSET = 255 + 10 + 40 + 35
FUTURE_OFFSET = 40


@pytest.fixture(scope="module")
def scheduled_calls(deploy_client, deployed_contracts, denoms, get_call):
    deploy_client.async_timeout = 30

    scheduler_contract = deployed_contracts.Scheduler
    client_contract = deployed_contracts.TestCallExecution

    anchor_block = deploy_client.get_block_number()

    blocks = (1, 4, 4, 8, 30, 40, 50, 60)

    calls = []

    for n in blocks:
        scheduling_txn = scheduler_contract.scheduleCall(
            client_contract._meta.address,
            client_contract.setBool.encoded_abi_signature,
            anchor_block + FUTURE_OFFSET + n,
            1000000,
            value=10 * denoms.ether,
            gas=3000000,
        )
        scheduling_receipt = deploy_client.wait_for_transaction(scheduling_txn)
        call = get_call(scheduling_txn)

        calls.append(call)
    return calls


@pytest.fixture(scope="session")
def mock_blocksage():
    class Mock(object):
        is_alive = True

    return Mock()


tests/v6/test_scheduler.py
import os
import pytest

from eth_alarm_client import (
    Scheduler,
    BlockSage,
)


V6_DIR = os.path.dirname(__file__)

project_dir = V6_DIR
deploy_contracts = [
    "Scheduler",
    "TestCallExecution",
]
deploy_client_type = "ipc"


@pytest.fixture(autouse=True)
def logging_config(monkeypatch):
    # Set to DEBUG for a better idea of what is going on in this test.
    monkeypatch.setenv('LOG_LEVEL', 'DEBUG')
    pass


FUTURE_OFFSET = 255 + 10 + 40 + 35


def test_scheduler(geth_node, deploy_client,
                   deployed_contracts, contracts,
                   get_call, denoms):
    deploy_client.async_timeout = 60
    block_sage = BlockSage(deploy_client)

    scheduler = deployed_contracts.Scheduler
    client_contract = deployed_contracts.TestCallExecution

    anchor_block = deploy_client.get_block_number()

    blocks = (1, 4, 4, 8, 30, 40, 50, 60)

    calls = []

    for n in blocks:
        scheduling_txn = scheduler.scheduleCall(
            client_contract._meta.address,
            client_contract.setBool.encoded_abi_signature,
            anchor_block + FUTURE_OFFSET + n,
            1000000,
            value=10 * denoms.ether,
            gas=3000000,
        )
        scheduling_receipt = deploy_client.wait_for_transaction(scheduling_txn)
        call = get_call(scheduling_txn)

        calls.append(call)

    for idx, call in enumerate(calls):
        print "Call #{0} should execute at block #{1}".format(idx, call.targetBlock())

    scheduler = Scheduler(scheduler, block_sage=block_sage)
    scheduler.monitor_async()

    last_call = calls[-1]
    final_block = last_call.targetBlock() + last_call.gracePeriod() + 1
    deploy_client.wait_for_block(
        final_block,
        2 * block_sage.estimated_time_to_block(final_block),
    )

    scheduler.stop()
    block_sage.stop()

    was_called = [call.wasCalled() for call in calls]
    assert all(was_called)


eth_alarm_client/cli.py
import time

import click

from eth_rpc_client import Client as RPCClient
from eth_ipc_client import Client as IPCClient
from eth_ipc_client.utils import get_default_ipc_path

from populus.contracts import Contract
from populus.contracts.common import EmptyDataError

from eth_alarm_client import (
    BlockSage,
    Scheduler,
)
from eth_alarm_client.contracts import contract_json


DEFAULT_ADDRESS = '0x6c8f2a135f6ed072de4503bd7c4999a1a17f824b'


def get_contract(contract_name):
    return Contract(contract_json[contract_name], contract_name)


@click.group()
def main():
    pass


@main.command()
@click.option(
    '--address',
    '-a',
    default=DEFAULT_ADDRESS,
    help="Runs the call execution scheduler",
)
@click.option(
    '--client',
    '-c',
    default='rpc',
    type=click.Choice(['rpc', 'ipc']),
    help="Whether the RPC or IPC client should be used.",
)
@click.option(
    '--rpchost',
    '-r',
    default='127.0.0.1',
    help="The RPC Host",
)
@click.option(
    '--rpcport',
    '-p',
    default='8545',
    help="The RPC Port",
)
@click.option(
    '--ipcpath',
    '-i',
    default=get_default_ipc_path,
    type=click.Path(exists=True, dir_okay=False),
    help="The RPC Port",
)
def scheduler(address, client, rpchost, rpcport, ipcpath):
    """
    Run the call scheduler.
    """
    if client == 'ipc':
        blockchain_client = IPCClient(ipc_path=ipcpath)
    elif client == 'rpc':
        blockchain_client = RPCClient(host=rpchost, port=rpcport)

    SchedulerContract = get_contract('Scheduler')

    scheduler_contract = SchedulerContract(address, blockchain_client)
    try:
        api_version = scheduler_contract.callAPIVersion()
        if api_version != 7:
            raise click.ClickException(
                "The scheduling contract address does not appear to have a compatable API"
            )
    except EmptyDataError:
        raise click.ClickException(
            "The scheduler address seems to not be correct.  Using {0}.  You "
            "may need to specify the address using `--address` if you are "
            "running the client against a test network".format(address)
        )

    block_sage = BlockSage(blockchain_client)
    scheduler = Scheduler(scheduler_contract, block_sage=block_sage)

    scheduler.monitor_async()

    try:
        while scheduler._thread.is_alive():
            time.sleep(1)

    except KeyboardInterrupt:
        scheduler.stop()
        scheduler.block_sage.stop()
        for scheduled_call in scheduler.active_calls.values():
            scheduled_call.stop()
        scheduler._thread.join(5)


if __name__ == '__main__':
    main()


eth_alarm_client/utils.py
import os
import logging
import collections
from logging import handlers


class cached_property(object):
    """
    Decorator that converts a method with a single self argument into a
    property cached on the instance.
    Optional ``name`` argument allows you to make cached properties of other
    methods. (e.g.  url = cached_property(get_absolute_url, name='url') )
    """
    def __init__(self, func, name=None):
        self.func = func
        self.__doc__ = getattr(func, '__doc__')
        self.name = name or func.__name__

    def __get__(self, instance, type=None):
        if instance is None:
            return self
        res = instance.__dict__[self.name] = self.func(instance)
        return res


class empty(object):
    pass


class _cache_once(object):
    """
    Similar to cached property except that it doesn't cache the value until it
    differs from the default value.
    """
    _cache_value = empty

    def __init__(self, func):
        self.func = func
        self.__doc__ = getattr(func, '__doc__')
        self.name = func.__name__

    def __get__(self, instance, type=None):
        value = self.func(instance)

        if value != self.default_value:
            instance.logger.debug("Caching return value: %s for function: %s", value, self.name)
            instance.__dict__[self.name] = value
        return value


def cache_once(default_value):
    return type('cache_once', (_cache_once,), {'default_value': default_value})


LEVELS = collections.defaultdict(lambda: logging.INFO)
LEVELS.update({
    'CRITICAL': logging.CRITICAL,
    'ERROR': logging.ERROR,
    'WARNING': logging.WARNING,
    'INFO': logging.INFO,
    'DEBUG': logging.DEBUG,
})


def get_logger(name, level=None):
    if level is None:
        level = LEVELS[os.environ.get('LOG_LEVEL', logging.INFO)]
    logger = logging.getLogger(name)
    logger.setLevel(logging.DEBUG)

    has_stream_handler = any(
        isinstance(handler, logging.StreamHandler) for handler in logger.handlers
    )
    if not has_stream_handler:
        stream_handler = logging.StreamHandler()
        stream_handler.setLevel(level)
        stream_handler.setFormatter(
            logging.Formatter(name.upper() + ': %(levelname)s: %(asctime)s %(message)s')
        )
        logger.addHandler(stream_handler)

    has_file_handler = any(
        isinstance(handler, handlers.RotatingFileHandler) for handler in logger.handlers
    )
    if not has_file_handler:
        if not os.path.exists('logs'):
            os.mkdir('logs')
        file_handler = handlers.RotatingFileHandler('logs/{0}.log'.format(name), maxBytes=10000000)
        file_handler.setLevel(logging.DEBUG)
        file_handler.setFormatter(logging.Formatter('%(levelname)s: %(asctime)s %(message)s'))
        logger.addHandler(file_handler)

    return logger


eth_alarm_client/scheduler.py
import threading
import time
import random

from ethereum.utils import denoms as denoms

from .block_sage import BlockSage
from .call_contract import CallContract
from .contracts import FutureBlockCall
from .utils import (
    get_logger,
    cached_property,
)


EMPTY_ADDRESS = '0x0000000000000000000000000000000000000000'


class Scheduler(object):
    _block_sage = None

    def __init__(self, scheduler, block_sage=None):
        self.logger = get_logger('scheduler')
        self.scheduler = scheduler

        if block_sage is None:
            block_sage = BlockSage(self.blockchain_client)
        self._block_sage = block_sage

        self.active_calls = {}
        self.active_claims = {}

    @property
    def block_sage(self):
        if self._block_sage is None:
            self.logger.error("Blocksage unexpectedly `None`")
            self._block_sage = BlockSage(self.blockchain_client)
        if not self._block_sage.is_alive:
            self.logger.error("Blocksage died.  Respawning")
            self._block_sage = BlockSage(self.blockchain_client)
        return self._block_sage

    @property
    def blockchain_client(self):
        return self.scheduler._meta.blockchain_client

    @cached_property
    def coinbase(self):
        return self.blockchain_client.get_coinbase()

    @property
    def is_alive(self):
        return self._thread.is_alive()

    @cached_property
    def minimum_grace_period(self):
        return self.scheduler.getMinimumGracePeriod()

    def monitor_async(self):
        self._run = True
        self._thread = threading.Thread(target=self.monitor)
        self._thread.daemon = True
        self._thread.start()

    def stop(self):
        self._run = False

    def monitor(self):
        while getattr(self, '_run', True):
            self.schedule_calls()
            self.claim_calls()
            self.cleanup_calls()
            self.cleanup_claim_threads()
            time.sleep(self.block_sage.block_time)

    def claim_calls(self):
        start_block = self.block_sage.current_block_number + 10 + 1
        end_block = self.block_sage.current_block_number + 255 + 10 - 1
        self.logger.debug("Looking for claimable calls between %s-%s", start_block, end_block)
        upcoming_calls = self.enumerate_calls(start_block, end_block)

        for call_address in upcoming_calls:
            if call_address in self.active_claims:
                self.logger.debug("Call %s already claimed", call_address)
                continue

            scheduled_call = CallContract(
                call_address=call_address,
                blockchain_client=self.blockchain_client,
                block_sage=self.block_sage,
            )

            if not scheduled_call.is_claimable:
                self.logger.debug("Call %s not claimable", call_address)
                continue

            current_balance = self.blockchain_client.get_balance(self.coinbase)
            if current_balance - 2 * scheduled_call.base_payment < 2 * denoms.ether:
                self.logger.error(
                    "Insufficient funds to claim %s.  Base Payment is %s ether",
                    scheduled_call.call_address,
                    scheduled_call.base_payment * 1.0 / denoms.ether,
                )
                continue

            fpcb = scheduled_call.first_profitable_claim_block
            if self.block_sage.current_block_number < fpcb:
                # claiming the call at this point would be committing to
                # execute it at a loss, and thus we will wait till at least the
                # maximum payment value for this call.
                self.logger.debug(
                    "Waiting till block %s to claim %s.  To claim before this block would be operating at a loss.",
                    fpcb,
                    scheduled_call.call_address,
                )
                continue

            # Random strategy.  Roll a number between 1-255.  If we are at
            # least this many blocks into the call window then claim the call.
            cbn = self.block_sage.current_block_number
            fcb = scheduled_call.first_claimable_block
            claim_block = cbn - fcb

            claim_if_above = random.randint(0, 255)

            self.logger.debug(
                "Claiming roll for %s, Rolled %s: Needed: %s",
                scheduled_call.call_address,
                claim_if_above,
                claim_block,
            )

            if claim_block > claim_if_above:
                # Asynchronously claim the call.  We don't want to wait for
                # these transactions since they could take a while and there
                # could be a lot of them.
                claim_thread = threading.Thread(target=self.claim_call, args=(scheduled_call,))
                claim_thread.daemon = True
                claim_thread.start()
                self.active_claims[call_address] = claim_thread

    def claim_call(self, scheduled_call):
        """
        Claim a call.
        """
        cbn = self.block_sage.current_block_number
        fcb = scheduled_call.first_claimable_block
        claim_block = cbn - fcb

        self.logger.info(
            "Attempting to claim call %s at block %s",
            scheduled_call.call_address,
            claim_block,
        )
        claim_txn = scheduled_call.claim()
        try:
            claim_receipt = self.blockchain_client.wait_for_transaction(
                claim_txn,
                10 * self.block_sage.block_time,
            )
            self.logger.info(
                "Call %s claimed with txn %s at claim block %s for %s ethers",
                scheduled_call.call_address,
                claim_txn,
                claim_block,
                scheduled_call.claim_amount * 1.0 / denoms.ether,
            )
            return claim_receipt
        except ValueError:
            # Handle timeout waiting for transaction.
            self.logger.error(
                "Timeout waiting for claim transaction %s for call %s",
                claim_txn,
                scheduled_call.call_address,
            )
            raise

    def schedule_calls(self):
        start_block = max(0, self.block_sage.current_block_number - self.minimum_grace_period)
        end_block = self.block_sage.current_block_number + 40
        self.logger.debug("Looking for calls between %s-%s", start_block, end_block)
        upcoming_calls = self.enumerate_calls(start_block, end_block)

        for call_address in upcoming_calls:
            self.logger.debug("Evaluating %s for scheduling", call_address)
            if call_address in self.active_calls:
                self.logger.debug("Call %s already scheduled", call_address)
                continue

            scheduled_call = CallContract(
                call_address=call_address,
                blockchain_client=self.blockchain_client,
                block_sage=self.block_sage,
            )

            if not scheduled_call.is_callable:
                self.logger.debug("Call %s not callable", call_address)
                continue

            self.logger.info("Tracking call: %s", scheduled_call.call_address)
            scheduled_call.execute_async()
            self.active_calls[call_address] = scheduled_call

    def cleanup_calls(self):
        for call_address, scheduled_call in tuple(self.active_calls.items()):

            if scheduled_call.txn_hash:
                self.logger.info("Removing finished call: %s", call_address)
                self.active_calls.pop(call_address)
            elif scheduled_call.last_block < self.block_sage.current_block_number:
                scheduled_call.stop()
                self.logger.info("Removing expired call: %s", call_address)
                self.active_calls.pop(call_address)
            elif not scheduled_call._thread.is_alive():
                self.logger.info("Removing dead call: %s", call_address)
                self.active_calls.pop(call_address)

    def cleanup_claim_threads(self):
        keys = tuple(self.active_claims.keys())
        for key in keys:
            if not self.active_claims[key].is_alive():
                self.active_claims.pop(key)

    def get_next_call(self, block_number):
        next_call = self.scheduler.getNextCall(block_number)
        if next_call == EMPTY_ADDRESS:
            return None
        return next_call

    def get_next_call_sibling(self, call_address):
        next_call = self.scheduler.getNextCallSibling(call_address)
        if next_call == EMPTY_ADDRESS:
            return None
        return next_call

    def enumerate_calls(self, left_block, right_block):
        """
        Query the scheduler contract for any calls that should be executed during
        the next 40 block window.
        """
        call_address = self.get_next_call(left_block)

        while call_address is not None:
            call = FutureBlockCall(call_address, self.blockchain_client)

            if left_block <= call.targetBlock() <= right_block:
                yield call_address
            else:
                break

            call_address = self.get_next_call_sibling(call_address)


eth_alarm_client/call_contract.py
import threading
import time
import math

from .block_sage import BlockSage
from .utils import (
    cached_property,
    cache_once,
    get_logger
)

from .contracts import (
    FutureBlockCall,
    CallLib,
)


EMPTY_ADDRESS = '0x0000000000000000000000000000000000000000'
CANCELLATION_WINDOW = 8

MAX_CALL_OVERHEAD_GAS = 200000
DEFAULT_CALL_GAS = 1000000


class CallContract(object):
    """
    Abstraction to represent an upcoming function call.
    """
    txn_hash = None
    txn_receipt = None
    txn = None

    _block_sage = None

    def __init__(self, call_address, blockchain_client, block_sage=None):
        self.blockchain_client = blockchain_client
        self.call_address = call_address
        self.call = FutureBlockCall(call_address, self.blockchain_client)
        self.logger = get_logger('call-{0}'.format(self.call_address))

        if block_sage is None:
            block_sage = BlockSage(self.blockchain_client)

        self._block_sage = block_sage

    @property
    def block_sage(self):
        if self._block_sage is None:
            self.stop()
            self.logger.error("Blocksage unexpectedly `None`")
            raise ValueError("Blocksage unexpectedly `None`")
        if not self._block_sage.is_alive:
            self.stop()
            self.logger.error("Blocksage died.  Killing Call")
            raise ValueError("Blocksage died")
        return self._block_sage

    #
    # Execution Pre Requesites
    #
    @cached_property
    def first_claimable_block(self):
        return self.target_block - 10 - 255

    @cached_property
    def first_profitable_claim_block(self):
        claim_cost = self.CLAIM_GAS_COST * self.blockchain_client.get_gas_price()
        block_number = min(240, int(math.ceil(claim_cost * 1.0 * 240 / self.base_payment)))
        return self.first_claimable_block + block_number

    @cached_property
    def last_claimable_block(self):
        return self.target_block - 10

    @property
    def get_claim_value(self, block_number):
        if block_number < self.first_claimable_block:
            return 0
        if block_number > self.target_block - 10 - 15:
            return self.base_payment
        n = block_number - self.first_claimable_block
        return int(self.base_payment * 1.0 * n / 240)

    @property
    def is_claimable(self):
        if self.is_cancelled:
            return False
        if self.claimer is not None:
            return False
        if self.block_sage.current_block_number < self.first_claimable_block:
            return False
        if self.block_sage.current_block_number > self.last_claimable_block:
            return False
        return True

    @property
    def is_callable(self):
        if self.was_called:
            return False
        elif self.is_cancelled:
            return False
        elif self.is_expired:
            return False
        elif not self.scheduler_can_pay:
            return False
        else:
            return True

    @property
    def is_expired(self):
        return self.block_sage.current_block_number >= self.last_block

    @property
    def scheduler_can_pay(self):
        gas_cost = self.get_execution_gas() * self.blockchain_client.get_gas_price()
        max_payment = 2 * self.base_payment
        max_donation = 2 * self.base_donation
        call_value = self.call_value

        return gas_cost + max_payment + max_donation + call_value < self.balance

    def stop(self):
        self._run = False

    def execute(self):
        # Blocks until we are within 3 blocks of the call window.
        self.logger.info("Sleeping until %s", self.target_block - 2)
        self.wait_for_call_window()
        self.logger.info("Entering call loop")

        while getattr(self, '_run', True):
            if self.is_expired:
                self.logger.error("Call window expired")
                break

            if self.was_called:
                self.logger.info("Call has already been executed")
                break

            if not self.scheduler_can_pay:
                self.logger.warning("Scheduler cannot pay for the call")
                break

            next_block_number = self.block_sage.current_block_number + 1
            if not self.should_call_on_block(next_block_number):
                time.sleep(2)
                continue

            # Execute the transaction
            self.logger.info("Attempting to execute call")
            txn_hash = self.call.execute(gas=self.get_execution_gas())

            # Wait for the transaction receipt.
            try:
                self.logger.debug("Waiting for transaction: %s", txn_hash)
                txn_receipt = self.blockchain_client.wait_for_transaction(
                    txn_hash,
                    self.block_sage.estimated_time_to_block(self.last_block) * 2,
                )
            except ValueError:
                self.logger.error("Unable to get transaction receipt: %s", txn_hash)
                break
            else:
                self.logger.info("Transaction accepted.")
                self.txn_hash = txn_hash
                self.txn_receipt = txn_receipt
                self.txn = self.blockchain_client.get_transaction_by_hash(txn_hash)

                # Check the log data from the executing transaction and log it.
                execution_logs = CallLib(None, self.blockchain_client).CallExecuted.get_transaction_logs(txn_hash)
                execution_data = tuple((
                    CallLib(None, self.blockchain_client).CallExecuted.get_log_data(log) for log in execution_logs
                ))
                abort_logs = CallLib(None, self.blockchain_client).CallAborted.get_transaction_logs(txn_hash)
                abort_data = tuple((
                    CallLib(None, self.blockchain_client).CallAborted.get_log_data(log) for log in abort_logs
                ))

                for entry in execution_data:
                    self.logger.info("Event:CallExecuted: %s", str(entry))
                for entry in abort_data:
                    self.logger.warning("Event:CallAborted: %s", str(entry))
                break

    def execute_async(self):
        self._run = True
        self._thread = threading.Thread(target=self.execute)
        self._thread.daemon = True
        self._thread.start()

    def wait_for_call_window(self, buffer=2):
        """
        wait for self.target_block - buffer (~30 seconds at 2 blocks)
        """
        if self.block_sage.current_block_number > self.last_block:
            raise ValueError("Already passed call execution window")

        while True:
            is_killed = not getattr(self, '_run', True)
            is_before_buffer = (
                self.block_sage.current_block_number < self.target_block - buffer
            )
            if not is_killed and is_before_buffer:
                time.sleep(
                    self.block_sage.estimated_time_to_block(
                        self.target_block - buffer,
                    ),
                )
            else:
                break

    #
    #  Meta Properties
    #
    @cached_property
    def coinbase(self):
        return self.blockchain_client.get_coinbase()

    @property
    def balance(self):
        """
        The account balance of the scheduler for this call.
        """
        return self.blockchain_client.get_balance(self.call_address)

    @cached_property
    def last_block(self):
        """
        The last block number that this call can be executed on.
        """
        return self.target_block + self.grace_period

    @cached_property
    def is_designated(self):
        return self.call.claimer != EMPTY_ADDRESS

    def can_call_at_block(self, block_number):
        return self.call.checkExecutionAuthorization(self.coinbase, block_number)

    def should_call_on_block(self, block_number):
        """
        Return whether an attempt to execute this call should be made on the
        provided block number.
        """
        # Before call window starts
        if self.target_block > block_number:
            return False

        # After call window ends.
        if block_number > self.last_block:
            return False

        if not self.is_designated:
            return True

        return self.can_call_at_block(block_number)

    def get_execution_gas(self):
        gas_limit = int(self.block_sage.current_block['gasLimit'], 16)
        return min(gas_limit, self.required_gas + 100000)

    #
    #  Call properties.
    #
    @cached_property
    def contract_address(self):
        return self.call.contractAddress()

    @cached_property
    def scheduler_address(self):
        return self.call.schedulerAddress()

    # The amount of gas to send with the claiming transaction.
    CLAIM_GAS = 500000

    # The cost in gas to claim the call.
    CLAIM_GAS_COST = 100000

    def claim(self, **kwargs):
        kwargs.setdefault('value', 2 * self.base_payment)
        kwargs.setdefault('gas', self.CLAIM_GAS)
        return self.call.claim(**kwargs)

    @cache_once(None)
    def claimer(self):
        _claimer = self.call.claimer()
        if _claimer == '0x0000000000000000000000000000000000000000':
            return None
        return _claimer

    @cache_once(0)
    def claimer_deposit(self):
        return self.call.claimerDeposit()

    @cache_once(0)
    def claim_amount(self):
        return self.call.claimAmount()

    @cached_property
    def target_block(self):
        return self.call.targetBlock()

    @cached_property
    def grace_period(self):
        return self.call.gracePeriod()

    @cached_property
    def call_value(self):
        return self.call.callValue()

    @cached_property
    def anchor_gas_price(self):
        return self.call.anchorGasPrice()

    @cached_property
    def required_gas(self):
        return self.call.requiredGas()

    @cached_property
    def required_stack_depth(self):
        return self.call.requiredStackDepth()

    @cached_property
    def base_payment(self):
        return self.call.basePayment()

    @cached_property
    def base_donation(self):
        return self.call.baseDonation()

    @cached_property
    def abi_signature(self):
        return self.call.abiSignature()

    @cache_once(False)
    def was_successful(self):
        return self.call.wasSuccessful()

    @cache_once(False)
    def was_called(self):
        return self.call.wasCalled()

    @cache_once(False)
    def is_cancelled(self):
        return self.call.isCancelled()
